### ISSUE TITLE
Editor graph representation

### ISSUE THREAD

**Issue by ilargierdi** [2graphic/sinap-ide#103]
### TL; DR: Should `GraphDelegate` have factory methods for creating nodes and edges?

As discussed earlier today, we agreed to give drawables a default implementation as opposed to being interfaces. The `DrawableGraph` was proposed to have its constructor take a `GraphDelegate` that would supply at the very least a validating method for edges. I do like the name `isValidEdge` better than `canCreateEdge`, but that is just a minor detail.

What my main concern for opening this discussion is what other methods to put on the `GraphDelegate` interface. You mentioned that it needs to have some notion of `activeNode` and `activeEdge` for creating nodes and edges. I'm not sure that the `DrawableGraph` should implement the creation logic for anything other than having wrapper methods to call something on the `GraphDelegate` for both validation and creation. The `GraphDelegate` would then be a factory and validator for graph elements in this case.

The issue I have with pushing graph element creation onto the `DrawableGraph` is that the class will need prior knowledge about `kind`s of nodes and edges and how to set their properties. This boils down to type casing on nodes and edges, which would have to be supplied by a third party, anyway.

The only concern the graph editor should have is drawing behavior, which are supplied by drawable properties. This was the main idea behind `DrawableEdge` and `DrawableNode`; they are literally just objects with properties and no functionality from the point of view of the editor. The editor does, however, need to be able to query something for new graph elements; it just doesn't need to know how to create them. I liked your proposal for a `GraphDelegate` interface, and I also like the idea of a factory object to create nodes and edges. So my question is: would it be reasonable to include factory methods on `GraphDelegate`?

**Comment by Sheyne** [comment]
You're right. The graph editor doesn't need any notion of an active node/edge type. Since that's the case, maybe we should just take a single `validator: (Node, Node) => boolean` argument.

**Comment by ilargierdi** [comment]
The editor still needs to be able to request new elements, though. I think supplying a `GraphDelegate` or `GraphFactory` object would be better. Maybe the factory methods could take a drawable as its parameter or maybe they take no arguments and return a drawable. Otherwise, the editor component could emit events to indicate when a graph element should be created or destroyed and the third party hooks into that, but that doesn't feel right.

**Comment by ilargierdi** [comment]
Or we could do something like Angular2's dependency injection, but that seems like it might take more effort.

**Comment by Sheyne** [comment]
So we've got a couple of things to tackle:

1. graph editor creates nodes and edge
2. graph editor wants to validate that an edge is valid
3. consumer requests nodes or edges

For 1 + 2, I think it's reasonable to have event listeners. These wouldn't need to be on a factory or delegate. They can follow the traditional HTML approach that an event listener can return false if it wants to cancel the event (in this case not create the node or edge). Maybe we even add an explicit validator to address 2. 

For 3, I think that the consumer (that being Sinap-IDE) should be able to get and set the current graph object on the editor. This allows us to implement tabs. This graph object should also expose methods to create nodes+edges. 

What do you think?

**Comment by ilargierdi** [comment]
Going off your last point, I definitely agree that the graph editor should require some notion of a graph object. I think the way it currently works should remain the same. Currently, the editor relies on this object for element creation and validation.

I'm not quite sure I understand how you envision your first point. For clarification, what you're saying is that the editor keeps track of its own internal representation of a graph; a third party will supply some kind of graph context to the editor that contains things like a reference to the editor representation along with zoom, pan, and selected items for context switching between tabs in addition to a validation method; the editor has some public field that gives third parties access to the current editor representation (which doesn't make sense to me if the context already has it); and, the editor will emit events for graph element creation and deletion in addition to the selection changes. Am I understanding correctly? If that's right, I think the only thing this changes with `DrawableGraph` is that it no longer has `(create|remove)(Edge|Node)`, and the editor would just throw events whenever it attempts to create or destroy a node or edge.

I think emitting events for node/edge creation/deletion is reasonable as you describe it, but if we use `EventEmitter`, it'll probably have to be synchronous to check the `cancelled` state if you want to use the event to prevent node/edge creation by the editor.

**Comment by ilargierdi** [comment]
To expand on my previous post, here's what I'm thinking:

```typescript
...

export class GraphEditorComponent {

    ...

    // Input for graph contextual stuff.
    @Input("graph")
    set setGraph(value: DrawableGraph | null) {
        // Unhook events from previous DrawableGraph.
        // Initialize internal graph representation from value.
        // Update canvas.
        // Note:
        // value is null when no tabs are open. Input events get unregistered
        // and the canvas gets cleared to 100% transparency.
    }

    ...

    @Output()
    selectionChanged: EventEmitter;

    // Called before a new node is displayed in the editor.
    @Output()
    creatingNode: EventEmitter; // Synchronous, probably.

    // Same as above for edges.
    @Output()
    creatingEdge: EventEmitter;

    ...

    // Some function {

        ...

        let drawable = new DrawableNode();
        let eventArgs = ... // { drawableProperties: Drawable, isCancelled ...}
        this.creatingNode.emit(eventArgs);
        // Handle cancelled.
        // Hook to drawable changes.
        // Update canvas.

        ...

    //}

    ...

}

...
```

```typescript
...

export class DrawableGraph {

    constructor(private validator: sometype) { }

    // Only contains properties.
    // Does not contain create/remove methods.

    ...

    isValidEdge(params) {
        return this.validator.isValidEdge(params);
    }

    ...

}

// Only contains properties and event emitters for property changed events.
// Does not use @Output for emitters. Maybe uses some other mechanism in
// Angular2 to signal changes that can be hooked by the editor inline. Or maybe
// uses our own change detection mechanism.
export class DrawableNode ...
export class DrawableEdge ...

...
```

Third parties use `DrawableGraph` by constructing it with a `validator` function/object, then bind it to `@Input("graph")`. This way the tab context can keep its reference to `DrawableGraph` and anything else it needs to know about, such as its associated plugin. The problem with this is that only the graph editor is allowed to create nodes and edges, since I don't think it's possible to listen for changes to the contents of a Set (e.g. `DrawableGraph.nodes`). We could move the event emitters to `DrawableGraph`, but that poses the same problem as the event emitters for `DrawableNode` and `DrawableEdge`. I think it's reasonable to only allow the editor to attempt to create nodes and edges; it will just have to be documented as an invariant on the nodes and edges sets.

**Comment by ilargierdi** [comment]
So I've been thinking about this issue coming from how the editor is implemented in a2fc10c. To give you a brief summary, the editor already has its own internal representation of a graph using wrapper classes around the `Drawable` interfaces. These wrapper classes, `GraphEditorNode` et al, know about its geometry and how to draw itself. This behavior is updated dynamically each time the editor makes a change to any of its properties. The problem is that the wrapper classes don't know how to listen for changes on its associated drawable interface.

Given this implementation, we can move the internals of the wrapper class directly into drawables, since they will be classes instead of interfaces, and each drawable can figure out how to update its geometry and draw behavior any time any one of its properties is changed from anywhere. The editor then doesn't have to know when a change was made, we can just revert back to calling `redraw` on on the editor.

The updated editor could behave just like my previous post where the editor receives some graph context stuff and raises events when attempting to create nodes or edges. The created drawable object will be passed as part of the event arguments, and listeners can do whatever they want with the drawable once they receive it.

**Comment by Sheyne** [comment]
I agree with everything you've said so far (I think, I'm only about 90% sure I understand) except that I think it's reasonable for `DrawableGraph` (now a class not an interface) to expose `createNode`, `createEdge`, `deleteNode`, and `deleteEdge`. `DrawableGraph` could then mark those edges/nodes as dirty for the next time the `DrawableGraph` got set to the editor, or if it was currently active in the editor, then inform the editor to run the necessary updates.

**Comment by ilargierdi** [comment]
With everything discussed so far, this is what I'm thinking:

```typescript
export class DrawableGraph {

    constructor(public readonly validator) { }

    // Same as before.
    @Input("graph")
    set setGraph(...) { ... }

    // Calling this will attempt to create a node. The creatingNode event will
    // be triggered here followed by the createdNode event if the event payoad
    // of the creatingNode event does not have isCancelled set.
    createNode(/* no params */) { ... }

    // This event is fired right before the node is created. The event payload
    // will contain a reference to the Drawable object so that its properties
    // can be set by any listener. Note that the collection of listeners will be
    // a set and each one will be called synchronously, so execution sequence is
    // not guaranteed to be the same for every event and the event payload will
    // not be threadsafe (for now, probably). The main purpose of this event is
    // to allow third parties to capture the Drawable object for future
    // reference.
    addCreatingNodeListener(listener) { ... }
    removeCreatingNodeListener(listener) { ... }

    // This event is fired right after the node is added to the DrawableGraph
    // set of nodes. Its event payload will contain the same Drawable object.
    // The main purpose of this event is to signal the editor to update the
    // canvas.
    addCreatedNodeListener(listener) { ... }
    removeCreatedNodeListener(listener) { ... }

    // Same stuff for edges

    ...

    // ... remove, create, validate, etc. ...

}

...

export class DrawableNode {

    ...

    // ... Display properties ...

    ...

    // This function is set internally by the object depending on what its
    // properties' values are.
    get draw: () => void;

    // Same as above.
    get update: () => void;

    ...

}

// Same for DrawableEdge
```

Note that nothing is passed to the `createNode` method. It will likely return the same Drawable object that is passed around the event listeners. With this, the editor can listen for `created` events and update itself accordingly regardless of where the call to `create*` was made. In the case that the editor is not currently displaying a `DrawableGraph` that is being abused with calls to `create*`, the editor won't care, because it will refresh the canvas as soon as `@Input("graph")` is changed.

Also, I'm only mentioning `create`, but assume the same or similar behavior for `delete`.

**Comment by Sheyne** [comment]
Looks good.

**Comment by ilargierdi** [comment]
I'll probably want to make these changes in the `graph-editor-maintenance` branch, since the majority of the groundwork has been laid out for this in there.

**Comment by Sheyne** [comment]
Why `[add/remove]CreatingNodeListener` rather than just an emitter?

**Comment by ilargierdi** [comment]
It's to allow anyone to subscribe to the emitter. The emitter gets called in `createNode`, but you subscribe to that event with `addEventListener`. Here, the `Event` is `CreatingNode`.

**Comment by ilargierdi** [comment]
Actually, the terminology I'm using here isn't very clear. My apologies. `createNode` will evoke each `CreatingNodeListener` right after it creates a new drawable, then evoke each `CreatedNodeListener` after the drawable has been added to `DrawableGraph.nodes`. The argument of `add*Listener` will get added to some private collection of listeners so that they can be evoked at the appropriate time.

**Comment by Sheyne** [comment]
I think that I'll allow isValidEdge to have any of its arguments be null. Nulls will act as a wildcard. This will allow it to say that the active edge type is never valid with the node you are dragging from.

**Comment by ilargierdi** [comment]
This sounds very useful. Since anchor points show up on nodes when the user hovers over them, we can use `isValidEdge` to prevent that from happening if the active edge cannot be created on the hovered node.

**Comment by Sheyne** [comment]
That was the use case I had in mind. Would you rather custom functions so it's less opaque than signaling with null?

**Comment by ilargierdi** [comment]
I was thinking something like this

```typescript
/**
 * EdgeValidator  
 *   Determines whether or not an edge is valid for a given source and
 *   destination node.
 * 
 *   If `dst` is not specified, the validator should check if an edge can be
 *   created from the source node.
 * 
 *   If `like` is specified, a drawable edge with a matching type of `like`
 *   is checked against the given source and destination nodes.
 */
export type EdgeValidator = (
    src: DrawableNode | null,
    dst?: DrawableNode,
    like?: DrawableEdge
) => boolean;
```

which has already been coded.

**Comment by ilargierdi** [comment]
#120
